src/__Private/Visitor/FragmentInfo.hack

```diff
@@ -0,0 +1,59 @@
+namespace Slack\GraphQL\__Private;
+
+use namespace HH\Lib\C;
+use namespace Graphpinator\Parser;
+use type \Slack\GraphQL\__Private\Utils\Stack;
+
+final class FragmentInfo extends ASTVisitor {
+    private ?Parser\Operation\Operation $current_operation = null;
+    private ?Parser\Fragment\Fragment $current_fragment = null;
+
+    private dict<string, Parser\Fragment\Fragment> $fragments = dict[];
+    private dict<int, vec<Parser\FragmentSpread\NamedFragmentSpread>> $operation_to_named_spread = dict[];
+    private dict<int, vec<Parser\FragmentSpread\NamedFragmentSpread>> $fragment_to_named_spread = dict[];
+
+    <<__Memoize>>
+    public function getFragments(Parser\Operation\Operation $operation): dict<string, Parser\Fragment\Fragment> {
+        $fragments = dict[];
+        $stack = new Stack($this->operation_to_named_spread[$operation->getId()] ?? vec[]);
+        while ($stack->length() > 0) {
+            $fragment = $this->fragments[$stack->pop()->getName()] ?? null;
+            if ($fragment && !C\contains_key($fragments, $fragment->getName())) {
+                $fragments[$fragment->getName()] = $fragment;
+                foreach ($this->fragment_to_named_spread[$fragment->getId()] ?? vec[] as $frag) {
+                    $stack->push($frag);
+                }
+            }
```

Comment:
This loop is doing some repetitive work that could probably be memoized.

Reply:
Any reason not to use recursion here instead of manually dealing with a stack? You could then put `<<__Memoize>>` on the recursive function (might have to be careful about infinite loops though).

Reply:
> might have to be careful about infinite loops though

This was the reason. It's doable using an instance variable though, or by passing a set of visited frags through the call stack.